text/0119-dynamic-component-creation.md

```diff
@@ -0,0 +1,415 @@
+
+Declaratively:
+
+```html
+<template>
+    <lwc:component lwc:is={ctor} class="container" onclick={handleClick}></lwc:component>
+</template>
+```
+
+Imperatively:
+
+```html
+<template>
+    <lwc:component lwc:is={lazyConstructor} lwc:ref="foo"></lwc:component>
+</template>
+```
+
+```javascript
+import { LightningElement } from "lwc";
+
+export default class extends LightningElement {
+    lazyConstructor;
+
+    connectedCallback() {
+        import('lightning/concreteComponent')
+            .then(({ default: ctor }) => this.lazyConstructor = ctor)
+            .catch(err => console.log('Error importing component'));
+    }
+
+    renderedCallback() {
+        if (this.refs.foo) {
+            this.refs.foo.setAttribute('title', 'hawaii');
+            this.refs.foo.addEventListener('click', () => console.log('hello world!'));
+        }
+    }
+}
+```
+#### Usage with other LWC directives
+
+All [LWC custom element directives](https://lwc.dev/guide/reference#directives-for-nested-templates) with the exception of `lwc:external` are available to use on `<lwc:component>` and their functionality will be passed through to the custom element once it has been created.
+
+`lwc:external` will not be available because the constructor provided to the `lwc:is` directive must be a `LightningElementConstructor`.
+
+#### Children of the dynamically created element
+
+The `<lwc:component>` tag accepts child elements, first creating the dynamic component and subsequently the children afterwards.  Each time the dynamic component changes, the existing dynamic component will be unmounted along with all of its children.  The new dynamic component will be created along with the children.
+
+```html
+<template>
+    <lwc:component lwc:is={ctor}>
+        <span>child</span>
+    </lwc:component>
+</template>
+```
+
+#### Light DOM vs shadow DOM
+
+A dynamic component can switch between light DOM and shadow DOM and should follow the same semantics as currently exists with other forms of dynamic rendering components.
+
+For example
+
+```html
+<template>
+    <template lwc:if={renderLight}>
+        <light-component></light-component>
+    <template lwc:else>
+        <shadow-component></shadow-component>
+    </template>
+</template>
+```
+
+Is semantically the same as
+
+```html
+<template>
+    <!-- The ctor swaps between light and shadow DOM components -->
+    <lwc:component lwc:is={ctor}></lwc:component>
+</template>
+```
+
+In both cases the dynamic component will be fully mounted and unmounted.
+
+#### Styles
+
+Something to keep in mind is that [light DOM styles are injected to the closest root node](https://rfcs.lwc.dev/rfcs/lwc/0115-light-dom#styles) and are not removed once the components are unmounted. This means that repeated mounting and mounting of a dynamic component could cause styles to be overwritten.  This can be avoided by using either scoped styles or shadow dom, as the styles will be scoped.
```

Comment:
this poses a new question not necessary related to this RFC, what happen if I have two instances of the same light component on the same template, and one of them is removed?

Reply:
I think in that case the styles will be inserted at the top of the root node.  If either or both of the elements are removed from the template, the styles will still be at the root node.

Reply:
I also noticed this reads incorrectly, I mean to say that repeated mounting and unmounting of different dynamic components could cause styles to be overwritten.

I'll update this section too.

Reply:
this is probably a question for @nolanlawson , how can we track down the multiple instances that contribute to a particular css style to be added?

Reply:
This section is incorrect:

> This means that repeated mounting and mounting of a dynamic component could cause styles to be overwritten

Styles are never overwritten in LWC. We simply append styles as needed. Two instances of the same light DOM component will result in one stylesheet being added to the nearest shadow root. This stylesheet is never removed (unless the shadow root itself is removed).

> how can we track down the multiple instances that contribute to a particular css style to be added?

@caridy Currently there is no debug information to figure this out; you have to just know which stylesheet came from which component.